File: python_sdk/src/reality_capture/specifications/geometry.py
from pydantic import BaseModel, Field
from typing import Optional


class BoundingBox(BaseModel):
    xmin: float = Field(description="X coordinate of the minimum corner")
    ymin: float = Field(description="Y coordinate of the minimum corner")
    zmin: float = Field(description="Z coordinate of the minimum corner")
    xmax: float = Field(description="X coordinate of the maximum corner")
    ymax: float = Field(description="Y coordinate of the maximum corner")
    zmax: float = Field(description="Z coordinate of the maximum corner")


class Point3d(BaseModel):
    x: float = Field(description="X coordinate")
    y: float = Field(description="Y coordinate")
    z: float = Field(description="Z coordinate")


class Coords2d(BaseModel):
    x: float = Field(description="X coordinate")
    y: float = Field(description="Y coordinate")


class Polygon2DWithHoles(BaseModel):
    outsideBounds: list[Coords2d] = Field(description="Outside bounds of the polygon")
    holes: Optional[list[list[Coords2d]]] = Field(default=None, description="List of holes boundaries if any")


class RegionOfInterest(BaseModel):
    crs: str = Field(description="Definition of the Region of Interest Coordinate System")
    polygons: list[Polygon2DWithHoles] = Field(description="List of polygons")
    altitude_min: float = Field(description="Minimum altitude", alias="altitudeMin")
    altitude_max: float = Field(description="Maximum altitude", alias="altitudeMax")

File: typescript/packages/reality-capture/src/specifications/geometry.ts
import { z } from "zod";

export const BoundingBoxSchema = z.object({
  xmin: z.number().describe("X coordinate of the minimum corner"),
  ymin: z.number().describe("Y coordinate of the minimum corner"),
  zmin: z.number().describe("Z coordinate of the minimum corner"),
  xmax: z.number().describe("X coordinate of the maximum corner"),
  ymax: z.number().describe("Y coordinate of the maximum corner"),
  zmax: z.number().describe("Z coordinate of the maximum corner"),
});
export type BoundingBox = z.infer<typeof BoundingBoxSchema>;

export const Point3dSchema = z.object({
  x: z.number().describe("X coordinate"),
  y: z.number().describe("Y coordinate"),
  z: z.number().describe("Z coordinate"),
});
export type Point3d = z.infer<typeof Point3dSchema>;

export const Coords2dSchema = z.object({
  x: z.number().describe("X coordinate"),
  y: z.number().describe("Y coordinate"),
});
export type Coords2d = z.infer<typeof Coords2dSchema>;

export const Polygon2DWithHolesSchema = z.object({
  outsideBounds: z.array(Coords2dSchema).describe("Outside bounds of the polygon"),
  holes: z.array(z.array(Coords2dSchema)).optional().describe("List of holes boundaries if any"),
});
export type Polygon2DWithHoles = z.infer<typeof Polygon2DWithHolesSchema>;

export const RegionOfInterestSchema = z.object({
  crs: z.string().describe("Definition of the Region of Interest Coordinate System"),
  polygons: z.array(Polygon2DWithHolesSchema).describe("List of polygons"),
  altitudeMin: z.number().describe("Minimum altitude"),
  altitudeMax: z.number().describe("Maximum altitude"),
});
export type RegionOfInterest = z.infer<typeof RegionOfInterestSchema>;